src/shim/runtime/v2/rune/agent/client/client.go

```diff
@@ -0,0 +1,187 @@
+// The codebase is inherited from kata-containers with the modifications.
+
+package client
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	grpcStatus "google.golang.org/grpc/status"
+
+	agentgrpc "github.com/confidential-containers/enclave-cc/src/shim/runtime/v2/rune/agent/grpc"
+	"github.com/containerd/ttrpc"
+)
+
+var defaultDialTimeout = 30 * time.Second
+
+var agentClientFields = logrus.Fields{
+	"name":   "agent-client",
+	"pid":    os.Getpid(),
+	"source": "agent-client",
+}
+
+const (
+	UnixSocketScheme = "unix"
+	TcpSocketScheme  = "tcp"
+)
+
+var agentClientLog = logrus.WithFields(agentClientFields)
+
+// AgentClient is an agent gRPC client connection wrapper for agentgrpc.AgentServiceClient
+type AgentClient struct {
+	ImageServiceClient agentgrpc.ImageService
+	conn               *ttrpc.Client
+}
+
+// NewAgentClient creates a new agent gRPC client and handles both unix and tcp addresses.
+//
+// Supported sock address formats are:
+//   - unix://<path>
+//   - tcp://<ip>:<port>
+func NewAgentClient(ctx context.Context, sock string, timeout uint32) (*AgentClient, error) {
+	grpcAddr, parsedAddr, err := parse(sock)
+	if err != nil {
+		return nil, err
+	}
+
+	dialTimeout := defaultDialTimeout
+	if timeout > 0 {
+		dialTimeout = time.Duration(timeout) * time.Second
+		agentClientLog.WithField("timeout", timeout).Debug("custom dialing timeout has been set")
+	}
+
+	var conn net.Conn
+	var d = agentDialer(parsedAddr)
+	conn, err = d(grpcAddr, dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	client := ttrpc.NewClient(conn)
+
+	return &AgentClient{
+		ImageServiceClient: agentgrpc.NewImageClient(client),
+		conn:               client,
+	}, nil
+}
+
+type dialer func(string, time.Duration) (net.Conn, error)
+
+// Close an existing connection to the agent gRPC server.
+func (c *AgentClient) Close() error {
+	return c.conn.Close()
+}
+
+func parse(sock string) (string, *url.URL, error) {
+	addr, err := url.Parse(sock)
+	if err != nil {
+		return "", nil, err
+	}
+
+	var grpcAddr string
+	// validate more
+	switch addr.Scheme {
```

Comment:
Supporting both `unix` and `tcp` looks like a lot of complexity especially because: 1) enclave-agent currently supports `tcp` only and 2) the mock server implemented in this PR only supports `unix`. Can we go with `tcp` (and have unit tests for it) and then do the switch.